File: day_12/12.py
import re


def calculate_price(ornaments: str) -> int:
    if not re.match(r"^[*o^#@]+$", ornaments):
        return None
    ornaments_values = {
        "*": 1,
        "o": 5,
        "^": 10,
        "#": 50,
        "@": 100,
    }
    acc = 0
    for index, ornament in enumerate(ornaments):
        ornament_value = (
            ornaments_values[ornaments[index + 1]] if index < len(ornaments) - 1 else 0
        )
        last_ornament_value = ornaments_values[ornament]
        if ornament_value > last_ornament_value:
            acc -= last_ornament_value
        else:
            acc += last_ornament_value
    return acc


print(calculate_price("***"))  # 3   (1 + 1 + 1)
print(calculate_price("*o"))  # 4   (5 - 1)
print(calculate_price("o*"))  # 6   (5 + 1)
print(calculate_price("*o*"))  # 5  (-1 + 5 + 1)
print(calculate_price("**o*"))  # 6  (1 - 1 + 5 + 1)
print(calculate_price("o***"))  # 8   (5 + 3)
print(calculate_price("*o@"))  # 94  (-5 - 1 + 100)
print(calculate_price("*#"))  # 49  (-1 + 50)
print(calculate_price("@@@"))  # 300 (100 + 100 + 100)
print(calculate_price("#@"))  # 50  (-50 + 100)
print(calculate_price("#@Z@"))  # None (Z es desconocido)

File: day_12/12.js
/** @param {string} ornaments
 * @return {number} - The price of the tree
 */
function calculatePrice(ornaments) {
  if (!/^[*o^#@]+$/g.test(ornaments)) return undefined;
  const ornamentsValues = {
    "*": 1,
    o: 5,
    "^": 10,
    "#": 50,
    "@": 100,
  };
  return Array.from(ornaments).reduce(
    (acc, ornament, index, ornamentsArray) => {
      const ornamentValue = ornamentsValues[ornamentsArray[index + 1]];
      const lastOrnamentValue = ornamentsValues[ornament];
      return ornamentValue > lastOrnamentValue
        ? acc - lastOrnamentValue
        : acc + lastOrnamentValue;
    },
    0
  );
}

console.log(calculatePrice("***")); // 3   (1 + 1 + 1)
console.log(calculatePrice("*o")); // 4   (5 - 1)
console.log(calculatePrice("o*")); // 6   (5 + 1)
console.log(calculatePrice("*o*")); // 5  (-1 + 5 + 1)
console.log(calculatePrice("**o*")); // 6  (1 - 1 + 5 + 1)
console.log(calculatePrice("o***")); // 8   (5 + 3)
console.log(calculatePrice("*o@")); // 94  (-5 - 1 + 100)
console.log(calculatePrice("*#")); // 49  (-1 + 50)
console.log(calculatePrice("@@@")); // 300 (100 + 100 + 100)
console.log(calculatePrice("#@")); // 50  (-50 + 100)
console.log(calculatePrice("#@Z@")); // undefined (Z es desconocido)